File: cv/cv-service/cv_service/tracking/objectstate.py
"""Pure-stdlib mirror of `cv_pb2.ObjectState` -- the wire-level snapshot of
internal tracking state (CV-ORCHESTRATION wave W1, plan §4.5).

**What this is, and is not.** Every field here has the same name, the same
group, and the same unit as its `cv_pb2.ObjectState` counterpart, verified
against `proto/vision/v1/cv.proto` field-by-field. It is deliberately NOT a
wrapper around the generated protobuf type: this module imports nothing from
`cv_service.gen` and nothing from `grpc`, so `orchestration/mirror.py` (the
only writer of these dataclasses) and every test that exercises it can run
with no proto stubs generated at all. `grpc/servicers.py` is the ONE place
that ever converts an `ObjectState` into a `cv_pb2.ObjectState` message --
the same sole-touchpoint discipline `tests/tracking/test_registry.py`
already greps for on every other `cv_pb2` symbol in this package, now
extended to this one.

**Absent group vs. zero field (plan §4.5's own distinction, load-bearing).**
`None` on a group (`identity`, `kinematics`, `belief`, `provenance`,
`memory`, `lock`, `timing`) means "this configuration does not compute this
at all" -- e.g. `lock` is `None` for every object on a stream that has never
had a lock applied, not `Lock(locked=False, ...)`. A GROUP that IS present
with a zero-valued FIELD inside it (`Memory(recovered=False,
identity_confidence=0.0, ...)` for a live track the gallery was asked about
but did not recover) is a genuine, meaningful zero, not an absence. Losing
this distinction on the wire would make "nobody asked" and "asked and found
nothing" indistinguishable, which is exactly the honesty gap
`docs/plans/active/CV-ORCHESTRATION-PLAN.md`'s ledger work (W0) already
closed for the debug surface -- W1 extends the same discipline to the wire.

**Enum values cross this module's boundary as their own value-name
strings**, mirroring `track.py`'s `STATE_*` / `engines/base.py`'s `SOURCE_*`
convention exactly: `OBJECT_LIFECYCLE_CONFIRMED == cv_pb2.ObjectLifecycle.
Name(cv_pb2.OBJECT_LIFECYCLE_CONFIRMED)`, verbatim, so `grpc/servicers.py`'s
mapping is a `cv_pb2.ObjectLifecycle.Value(state.lifecycle)` call and
nothing else -- no lookup table anywhere to drift out of sync with the
`.proto` file.

**Boxes** (`Kinematics.box`/`.detector_box`/`.tracker_box`/`.predicted_box`)
are `cv_service.tracking.engines.base.Box` (normalized `[0, 1]`, same units
as `cv_pb2.BoundingBox`) or `None` -- never a zero-valued `Box`, for the same
absent-vs-zero reason as the groups above: a track this frame never had a
detector-sourced box for reports `detector_box=None`, not a box collapsed to
the origin.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from cv_service.tracking.engines.base import Box

# Value-name strings, one-for-one with `cv_pb2.ObjectLifecycle`'s own enum
# value names (`track.py`'s `STATE_*` / `engines/base.py`'s `SOURCE_*`
# convention, extended here). Never imported from `cv_pb2` -- see this
# module's own docstring on why -- so if the `.proto` file's value names
# ever change, this list must be updated by hand alongside it; nothing
# machine-checks the match other than `grpc/servicers.py`'s own
# `cv_pb2.ObjectLifecycle.Value(...)` call raising at runtime.
OBJECT_LIFECYCLE_UNSPECIFIED = "OBJECT_LIFECYCLE_UNSPECIFIED"
OBJECT_LIFECYCLE_TENTATIVE = "OBJECT_LIFECYCLE_TENTATIVE"
OBJECT_LIFECYCLE_CONFIRMED = "OBJECT_LIFECYCLE_CONFIRMED"
OBJECT_LIFECYCLE_COASTING = "OBJECT_LIFECYCLE_COASTING"
OBJECT_LIFECYCLE_LOST = "OBJECT_LIFECYCLE_LOST"
OBJECT_LIFECYCLE_DORMANT = "OBJECT_LIFECYCLE_DORMANT"

# Value-name strings, one-for-one with `cv_pb2.EvidenceSource`. Same caveat
# as above.
EVIDENCE_SOURCE_UNSPECIFIED = "EVIDENCE_SOURCE_UNSPECIFIED"
EVIDENCE_SOURCE_DETECTOR = "EVIDENCE_SOURCE_DETECTOR"
EVIDENCE_SOURCE_TRACKER = "EVIDENCE_SOURCE_TRACKER"
EVIDENCE_SOURCE_PREDICTED = "EVIDENCE_SOURCE_PREDICTED"
EVIDENCE_SOURCE_MEMORY = "EVIDENCE_SOURCE_MEMORY"
EVIDENCE_SOURCE_REUPDATE = "EVIDENCE_SOURCE_REUPDATE"


@dataclass(frozen=True)
class LabelCandidate:
    """`ObjectState.Identity.LabelCandidate` -- one entry in the decayed vote
    tally `track.py`'s `_update_label_election` already computes every call
    (`Track.label_tally`), before this wave discarded it once the winner was
    picked.
    """

    label: str
    weight: float


@dataclass(frozen=True)
class Identity:
    """`ObjectState.Identity`. `label` is the ELECTED label
    (`Track.elected_label`); `label_raw` is the raw, unelected label this
    frame's own observation carried -- the same "stable name vs. one
    detector pass's own word" split `assign.py`'s candidate/target already
    draws.
    """

    label: str
    label_raw: str
    candidates: "tuple[LabelCandidate, ...]" = ()
    stability: int = 0


@dataclass(frozen=True)
class Kinematics:
    """`ObjectState.Kinematics`. `box` is the track's SETTLED box for this
    frame (`Track.box`, post-`_observe`); `detector_box`/`tracker_box`/
    `predicted_box` are the same "what did each SOURCE actually see"
    breakdown `session.py`/`outcome.py` already keep separate elsewhere,
    each `None` when this frame has no evidence of that kind for this
    object -- never a box collapsed to zero.
    """

    box: Optional[Box] = None
    detector_box: Optional[Box] = None
    tracker_box: Optional[Box] = None
    predicted_box: Optional[Box] = None
    horizon_ms: int = 0
    velocity_x: float = 0.0
    velocity_y: float = 0.0
    displacement_x: float = 0.0
    displacement_y: float = 0.0
    motion_compensated: bool = False


@dataclass(frozen=True)
class Belief:
    """`ObjectState.Belief`. `confidence_raw` is this frame's own observed
    confidence (`Track.confidence`); `confidence_smoothed` is the EMA
    (`Track.confidence_smoothed`, `track.py`'s `_CONFIDENCE_SMOOTHING`).
    """

    confidence_raw: float = 0.0
    confidence_smoothed: float = 0.0
    existence: float = 0.0
    since_confirmed_ms: int = 0


@dataclass(frozen=True)
class Provenance:
    """`ObjectState.Provenance`. `source` is one of the `EVIDENCE_SOURCE_*`
    constants above; `contributors` are contributor ids from this frame's
    own `FrameLedger.objects[track_id]` (read-only -- see `mirror.py`'s
    module docstring for why the ledger is read here and never mutated).
    """

    source: str = EVIDENCE_SOURCE_UNSPECIFIED
    contributors: "tuple[str, ...]" = ()
    assoc_cost: float = 0.0
    reupdated: bool = False


@dataclass(frozen=True)
class Memory:
    """`ObjectState.Memory`. Present (never `None`) for every LIVE object
    once a gallery is resolved for the stream at all, whether or not this
    particular object was recovered this particular frame -- see
    `mirror.py`'s module docstring on why that is a deliberate reading of
    plan §4.5's "absent group" rule, not a contradiction of it.
    """

    recovered: bool = False
    identity_confidence: float = 0.0
    dormant_ms: int = 0
    gallery_matches: int = 0
    match_distance: float = 1.0


@dataclass(frozen=True)
class Lock:
    """`ObjectState.Lock`. Present for every live object once `LockArbiter`
    has ever applied a lock on this stream (`applied_seq > 0`), `None`
    before that -- see `mirror.py`'s module docstring.
    """

    locked: bool = False
    lock_seq_applied: int = 0


@dataclass(frozen=True)
class Timing:
    """`ObjectState.Timing`. All four of `first_seen_ms`/`last_seen_ms`/
    `last_confirmed_ms` are MILLIS on the wire; `Track.first_seen`/
    `.last_seen`/`.last_confirmed` are monotonic SECONDS
    (`TrackBook.apply`'s own docstring), so `mirror.py` is responsible for
    the `* 1000.0` conversion on every live-track field -- `DormantIdentity`.
    lost_at_millis`/`.first_seen_millis` are already millis (`track.py`'s
    `_retire`), so a dormant entry's `Timing` needs no such conversion.
    """

    first_seen_ms: int = 0
    last_seen_ms: int = 0
    last_confirmed_ms: int = 0
    age_frames: int = 0
    hits: int = 0
    misses: int = 0


@dataclass(frozen=True)
class ObjectState:
    """`cv_pb2.ObjectState`'s pure-stdlib mirror -- see this module's own
    docstring for the absent-group/zero-field distinction, the box-is-
    `None`-not-zero rule, and the enum-as-value-name-string convention that
    govern every field below.

    `id`/`lifecycle`/`stream_id` are the only fields every entry (live or
    dormant) always populates; every group below them is `Optional` and
    `None` exactly when `orchestration/mirror.py` had nothing to compute for
    it this frame.
    """

    id: int
    lifecycle: str
    stream_id: str
    identity: Optional[Identity] = None
    kinematics: Optional[Kinematics] = None
    belief: Optional[Belief] = None
    provenance: Optional[Provenance] = None
    memory: Optional[Memory] = None
    lock: Optional[Lock] = None
    timing: Optional[Timing] = None

File: contexts/vision-perception/src/main/java/com/drones/vision/perception/domain/model/ObjectState.java
package com.drones.vision.perception.domain.model;

import com.drones.vision.kernel.BoundingBox;
import com.drones.vision.kernel.StreamId;
import java.util.List;

/**
 * One object, mirrored (docs/plans/active/CV-ORCHESTRATION-PLAN.md §4.5) — the platform's whole
 * current view of one tracked or dormant identity, grouped by facet instead of a box with facts
 * bolted on as flat scalars. Keyed by {@code (streamId, id)}; {@code id} is the same number as
 * {@link Detection}'s {@code TrackRef#trackId()}.
 *
 * <p><strong>A {@code null} nested group is not the forbidden "null means the feature is
 * off."</strong> It means this configuration does not <em>compute</em> those facts at all — e.g.
 * {@code kinematics.predictedBox()} is never populated in FOLLOW mode, where the book is never
 * whole-predicted, so the absence is honest, not a missing default. A field inside a
 * <em>present</em> group, by contrast, is always a genuine fact, including a genuine zero. There
 * is deliberately no {@code hasX()} ceremony here — a consumer checks the accessor for
 * {@code null} directly, same as every other nullable component in this package.
 *
 * <p>{@code memory}/{@code lock} are named {@link MemoryFacts}/{@link LockFacts} here, not the
 * wire's bare {@code Memory}/{@code Lock} — those two words as top-level Java type names in a
 * domain package would read as generic utility types, not "recovery facts about this object" /
 * "follow-lock facts about this object." The accessors stay {@code memory()}/{@code lock()} so
 * the JSON/wire field names this record mirrors are unchanged.
 *
 * @param id         per-stream track id, matching the {@link TrackRef#trackId()} of the {@link
 *                   Detection} this frame also carries for the same object; must be positive —
 *                   {@code 0} is the wire's untracked sentinel and must never reach the domain
 * @param lifecycle  where this identity is in its life; never {@code null}
 * @param streamId   stream this object belongs to; never {@code null}
 * @param identity   label and label-election facts, or {@code null} if not computed
 * @param kinematics box and motion facts, or {@code null} if not computed
 * @param belief     confidence/existence facts, or {@code null} if not computed
 * @param provenance which contributor(s) produced this frame's view of the object, or {@code
 *                   null} if not computed
 * @param memory     re-acquisition facts, or {@code null} if not computed
 * @param lock       FOLLOW-lock facts, or {@code null} if not computed
 * @param timing     the lifecycle counters death is decided from, or {@code null} if not computed
 */
public record ObjectState(long id, ObjectLifecycle lifecycle, StreamId streamId, Identity identity,
                           Kinematics kinematics, Belief belief, Provenance provenance, MemoryFacts memory,
                           LockFacts lock, Timing timing) {

    public ObjectState {
        if (id < 1) {
            throw new IllegalArgumentException(
                    "ObjectState id must be positive (0 is the wire's untracked sentinel): " + id);
        }
        if (lifecycle == null) {
            throw new IllegalArgumentException("ObjectState lifecycle must not be null");
        }
        if (streamId == null) {
            throw new IllegalArgumentException("ObjectState streamId must not be null");
        }
    }

    /**
     * One entry of the label-vote tally behind {@link Identity#label()} — the distribution, not
     * just the winner, so a training-capture consumer can audit an election it disagrees with.
     *
     * @param label  a candidate label; must not be {@code null}
     * @param weight decayed, confidence-weighted tally; comparable only within one frame; must be
     *               finite
     */
    public record LabelCandidate(String label, double weight) {
        public LabelCandidate {
            if (label == null) {
                throw new IllegalArgumentException("LabelCandidate label must not be null");
            }
            if (!Double.isFinite(weight)) {
                throw new IllegalArgumentException("LabelCandidate weight must be finite: " + weight);
            }
        }
    }

    /**
     * What the object is called, and how settled that name is.
     *
     * <p>{@code labelRaw} may be blank — a frame with no detector observation this frame (a pure
     * coast, or a recovery from memory) still carries the elected {@code label}, but has no raw
     * detector label of its own to report.
     *
     * @param label      elected label — hysteresis-gated, what an operator reads; must not be
     *                   {@code null}
     * @param labelRaw   this frame's own detector label, un-elected; must not be {@code null},
     *                   may be blank
     * @param candidates the tally behind the election, best-first; defensively copied, never
     *                   {@code null}
     * @param stability  consecutive elections since {@code label} last changed; must not be
     *                   negative
     */
    public record Identity(String label, String labelRaw, List<LabelCandidate> candidates, int stability) {
        public Identity {
            if (label == null) {
                throw new IllegalArgumentException("Identity label must not be null");
            }
            if (labelRaw == null) {
                throw new IllegalArgumentException("Identity labelRaw must not be null");
            }
            if (candidates == null) {
                throw new IllegalArgumentException("Identity candidates must not be null");
            }
            if (stability < 0) {
                throw new IllegalArgumentException("Identity stability must not be negative: " + stability);
            }
            candidates = List.copyOf(candidates);
        }
    }

    /**
     * Where the object is, and how that was arrived at. Three boxes are kept separate rather than
     * merged into one "current box" (the DeepStream separation, plan §4.5 R5): {@code box} is the
     * elected box — identical to the {@link Detection} box this frame also carries for the same
     * object — and the other three say what each source independently claimed, each {@code null}
     * when that source produced nothing this frame.
     *
     * @param box                the elected box; must not be {@code null}
     * @param detectorBox        what the detector claimed this frame, or {@code null} if none
     * @param trackerBox         what the single-object tracker claimed, or {@code null}; FOLLOW
     *                           only
     * @param predictedBox       constant-velocity extrapolation to this frame's instant, or
     *                           {@code null} if not computed
     * @param horizonMillis      how far {@code predictedBox} was extrapolated; must not be
     *                           negative
     * @param velocityX          normalized frame-widths per second, matching {@link
     *                           TrackRef#velocityX()}'s units; must be finite
     * @param velocityY          normalized frame-heights per second; must be finite
     * @param displacementX      box-centre delta since the previous update, normalized; must be
     *                           finite
     * @param displacementY      box-centre delta since the previous update, normalized; must be
     *                           finite
     * @param motionCompensated  whether a non-identity ego-motion transform was applied
     */
    public record Kinematics(BoundingBox box, BoundingBox detectorBox, BoundingBox trackerBox,
                              BoundingBox predictedBox, long horizonMillis, double velocityX, double velocityY,
                              double displacementX, double displacementY, boolean motionCompensated) {
        public Kinematics {
            if (box == null) {
                throw new IllegalArgumentException("Kinematics box must not be null");
            }
            if (horizonMillis < 0) {
                throw new IllegalArgumentException("Kinematics horizonMillis must not be negative: " + horizonMillis);
            }
            if (!Double.isFinite(velocityX)) {
                throw new IllegalArgumentException("Kinematics velocityX must be finite: " + velocityX);
            }
            if (!Double.isFinite(velocityY)) {
                throw new IllegalArgumentException("Kinematics velocityY must be finite: " + velocityY);
            }
            if (!Double.isFinite(displacementX)) {
                throw new IllegalArgumentException("Kinematics displacementX must be finite: " + displacementX);
            }
            if (!Double.isFinite(displacementY)) {
                throw new IllegalArgumentException("Kinematics displacementY must be finite: " + displacementY);
            }
        }
    }

    /**
     * How strongly the platform believes the object is there at all.
     *
     * @param confidenceRaw      this frame's own detector confidence, range [0,1]
     * @param confidenceSmoothed EMA across this identity's observations, range [0,1]
     * @param existence          [0,1] from hits/misses — survives a frame with no detection
     * @param sinceConfirmedMillis time since the last DETECTOR confirmation, not since birth;
     *                             must not be negative
     */
    public record Belief(double confidenceRaw, double confidenceSmoothed, double existence,
                          long sinceConfirmedMillis) {
        public Belief {
            requireUnitRange(confidenceRaw, "confidenceRaw");
            requireUnitRange(confidenceSmoothed, "confidenceSmoothed");
            requireUnitRange(existence, "existence");
            if (sinceConfirmedMillis < 0) {
                throw new IllegalArgumentException(
                        "Belief sinceConfirmedMillis must not be negative: " + sinceConfirmedMillis);
            }
        }
    }

    /**
     * Who said so, this frame. {@code contributors} is the ledger's own attribution, summarized —
     * the claims themselves stay in the ledger, never here (algorithm-shaped facts belong to the
     * ledger, not the mirror, plan §4.5 E7).
     *
     * @param source       what kind of evidence produced this frame's view; must not be {@code
     *                     null}
     * @param contributors contributor ids that made a claim about this object this frame;
     *                     defensively copied, never {@code null}
     * @param assocCost    the matched pair's total cost; {@code 0} when nothing matched; must not
     *                     be negative
     * @param reupdated    whether ORU reconstructed this track's gap this frame
     */
    public record Provenance(EvidenceSource source, List<String> contributors, double assocCost,
                              boolean reupdated) {
        public Provenance {
            if (source == null) {
                throw new IllegalArgumentException("Provenance source must not be null");
            }
            if (contributors == null) {
                throw new IllegalArgumentException("Provenance contributors must not be null");
            }
            if (assocCost < 0) {
                throw new IllegalArgumentException("Provenance assocCost must not be negative: " + assocCost);
            }
            contributors = List.copyOf(contributors);
        }
    }

    /**
     * Re-acquisition facts. Today's wire carries {@code identity_confidence}/{@code
     * dormant_millis} per detection, but the JSON republishes them only for the FOLLOW-locked
     * track (plan §4.5 D6); here they belong to the object itself, whoever is watching it.
     *
     * @param recovered          this frame re-acquired the identity from the gallery
     * @param identityConfidence the match score that justified it, range [0,1]
     * @param dormantMillis      how long it had been dormant — for a DORMANT object, still is;
     *                           must not be negative
     * @param galleryMatches     dormant identities considered before this answer; must not be
     *                           negative
     * @param matchDistance      best appearance distance across the gallery, range [0,1]
     */
    public record MemoryFacts(boolean recovered, double identityConfidence, long dormantMillis, int galleryMatches,
                               double matchDistance) {
        public MemoryFacts {
            requireUnitRange(identityConfidence, "identityConfidence");
            if (dormantMillis < 0) {
                throw new IllegalArgumentException(
                        "MemoryFacts dormantMillis must not be negative: " + dormantMillis);
            }
            if (galleryMatches < 0) {
                throw new IllegalArgumentException(
                        "MemoryFacts galleryMatches must not be negative: " + galleryMatches);
            }
            requireUnitRange(matchDistance, "matchDistance");
        }
    }

    /**
     * Whether FOLLOW is holding this object. The frame-level {@code locked_track_id} on the wire
     * says <em>which</em> object; this says so on the object itself, the only form a renderer can
     * use without a second lookup.
     *
     * @param locked          whether FOLLOW currently holds this object
     * @param lockSeqApplied  the lock generation actually in force; must not be negative
     */
    public record LockFacts(boolean locked, long lockSeqApplied) {
        public LockFacts {
            if (lockSeqApplied < 0) {
                throw new IllegalArgumentException(
                        "LockFacts lockSeqApplied must not be negative: " + lockSeqApplied);
            }
        }
    }

    /**
     * The counters death is decided from (plan §4.5 E4).
     *
     * <p><b>The three instants are not epoch time.</b> They share the timebase of the response's
     * own {@code timestamp_millis} and of no other clock: cv-service ages tracks on
     * {@code time.monotonic()} and rebases them once, at the wire, onto whatever timestamp that
     * frame carries. Under {@code DetectStream} that anchor is this JVM's clock (it stamped the
     * request), under {@code DetectPulled} it is the service host's. Compare them against
     * {@link DetectionResult#capturedAt()} of the result they arrived on, never against a reader's
     * wall clock. The durations elsewhere on this record family ({@code horizonMillis},
     * {@code sinceConfirmedMillis}, {@code dormantMillis}) carry no timebase at all and need no
     * such care.
     *
     * <p>The non-negative checks below therefore assume that anchor is a real epoch stamp, which
     * it is on every live path ({@code DetectionFrameCodec#encode} always stamps the frame's
     * capture time). A caller that sent {@code 0} would rebase these into the negative and this
     * constructor would reject that object — which is the intended outcome, and why the codec
     * drops the single offending object with a warning rather than letting it fail the frame.
     *
     * @param firstSeenMillis     when this identity was first observed, on the response's timebase;
     *                            must not be negative
     * @param lastSeenMillis      when this identity was last updated by any evidence, on the
     *                            response's timebase; must not be negative
     * @param lastConfirmedMillis when the detector last confirmed this identity, on the response's
     *                            timebase; must not be negative
     * @param ageFrames           frames since this identity was born; must not be negative
     * @param hits                confirming frames; must not be negative
     * @param misses              consecutive unmatched frames; must not be negative
     */
    public record Timing(long firstSeenMillis, long lastSeenMillis, long lastConfirmedMillis, int ageFrames,
                          int hits, int misses) {
        public Timing {
            if (firstSeenMillis < 0) {
                throw new IllegalArgumentException("Timing firstSeenMillis must not be negative: " + firstSeenMillis);
            }
            if (lastSeenMillis < 0) {
                throw new IllegalArgumentException("Timing lastSeenMillis must not be negative: " + lastSeenMillis);
            }
            if (lastConfirmedMillis < 0) {
                throw new IllegalArgumentException(
                        "Timing lastConfirmedMillis must not be negative: " + lastConfirmedMillis);
            }
            if (ageFrames < 0) {
                throw new IllegalArgumentException("Timing ageFrames must not be negative: " + ageFrames);
            }
            if (hits < 0) {
                throw new IllegalArgumentException("Timing hits must not be negative: " + hits);
            }
            if (misses < 0) {
                throw new IllegalArgumentException("Timing misses must not be negative: " + misses);
            }
        }
    }

    private static void requireUnitRange(double value, String name) {
        if (Double.isNaN(value) || value < 0.0 || value > 1.0) {
            throw new IllegalArgumentException("ObjectState " + name + " must be within [0,1]: " + value);
        }
    }
}
